File: 1106-parsing-a-boolean-expression/1106-parsing-a-boolean-expression.java
class Solution {
    public boolean parseBoolExpr(String expression) {
        if(expression.length() == 1) return expression.charAt(0) == 't';
        
        ArrayDeque<Character> stack = new ArrayDeque<>();
        
        for(char ch : expression.toCharArray())
        {
            if(ch == ')')
            {
                boolean isFalsePresent = false;
                boolean isTruePresent = false;
                char res = '/';
                
                while(!stack.isEmpty() && stack.peek() != '(')
                {
                    if(stack.pop() == 't') isTruePresent = true;
                    else isFalsePresent = true;
                }
                stack.pop();
                char operator = stack.pop();
                if(operator == '&') res = isFalsePresent ? 'f' : 't';
                else if(operator == '|') res = isTruePresent ? 't' : 'f';
                else if(operator == '!') res = isTruePresent ? 'f' : 't';
                stack.push(res);
            }
            else if(ch != ',')
                stack.push(ch);
        }
        return stack.peek() == 't';
    }
}

File: 1106-parsing-a-boolean-expression/1106-parsing-a-boolean-expression.py
class Solution:
    def parseBoolExpr(self, expression: str) -> bool:
        stack = []
        
        for ch in expression:
            if(ch == ')'):
                isTruePresent = False
                isFalsePresent = False
                
                while stack[-1] != '(':
                    if stack.pop() == 't': isTruePresent = True
                    else : isFalsePresent: isFalsePresent = True
                
                stack.pop()
                
                operator = stack.pop()
                res = None
                if operator == '&': res = 'f' if isFalsePresent else 't'
                elif operator == '|': res = 't' if isTruePresent else 'f'
                elif operator == '!': res = 'f' if isTruePresent else 't'
                    
                stack.append(res)
                
            elif(ch != ','):
                stack.append(ch)
                
        return stack[0] == 't'